Python/eof.py

```diff
@@ -0,0 +1,198 @@
+import numpy as np
+from sklearn.decomposition import pca 
+from sklearn import linear_model
+
+from operators import *
+
+def ensemble_reshape(data_cube):
+    """Reshapes data cube into array shape 
+    for eof calculations. 
+    --------------------------------
+    Args: 
+        data_cube (xarray Dataset): datacube 
+            containing an ensemble of 
+            CanRCM4 models 
+    Returns:
+        field_ens (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    """
+    check_keys(data_cube)
+
+    x_field = data_cube['dv']
+    n, p = cell_count(data_cube)
+    field_ens = np.reshape(x_field.values, (p, n))
+
+    return field_ens
+
+def get_land(ensemble_arr):
+    """Gets indeces of nan values to 
+    identify where land or ocean is. 
+    --------------------------------
+    Args: 
+        ensemble_arr (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    Returns:
+        (numpy.ndarray): 2-d array containing
+            the indeces of nan values
+    """
+    return np.isnan(ensemble_arr)
+
+def rand_sample_index(p, frac):
+    """Randomly sample a range of integers. 
+    --------------------------------
+    Args: 
+        p (int): maximum integer to sample
+        frac (float): fractional size of p 
+            to sample
+    Returns:
+        (numpy.ndarray): array containing indeces from 
+            a fraction of 0 to p
+    """
+    index = np.random.choice(np.arange(p), int(frac*p))
+    return index
+
+def pseudo_obs(data_cube, frac=0.02):
+    """Randomly sample a data cube to generate
+    pseudo observations from ensemble. 
+    --------------------------------
+    Args:  
+        data_cube (xarray Dataset): datacube 
+            containing an ensemble of 
+            CanRCM4 models 
+    Returns:
+        y_sample (numpy.ndarray): fraction
+            of random spatially sampled design
+            value grid cells
+        index (numpy.ndarray): index locations 
+            of y_sample from original ensemble size
+    """
+    Y_ens = ensemble_reshape(data_cube)
+    land_idx = get_land(Y_ens)
+    # select only land
+    Y_land = Y_ens[~land_idx.any(axis=1)]
```

Comment:
What exactly is this doing? Selecting values where there may or value in the ensemble is NA?